README.md
# GTNotification
An in-app notification banner for Swift.

**Setup with Cocoapods (recommended setup)**

- Add the instructions below to your Podfile:

```
source 'https://github.com/CocoaPods/Specs.git'
platform :ios, '9.0'
use_frameworks!

pod 'GTNotification', '0.1'
```

- Then add ```import GTNotification``` at the top of your Swift class.

- Finally compile and run and you will be good to go.

**Otherwise you can also manually import the framework**

Just drag and drop ```'GTNotification/*.swift``` file(s) into your Xcode project and do not forget to check the checkbox entitled "Copy items if needed" and the appropriate checkboxe(s) in "Add to targets" section.

**Example**

``` swift
// Initialize a notification
let notification: GTNotification = GTNotification()
notification.title = "Welcome Notification"
notification.message = "Thank you for checking out GTNotificationView."
notification.position = GTNotificationPosition.Top
notification.animation = GTNotificationAnimation.Slide
notification.blurEnabled = true

// Perform a custom selector on tap
notification.addTarget(self, action: Selector("dismissNotification"))

// Set the notification's delegate
notification.delegate = self

// Show the notification
GTNotificationManager.sharedInstance.showNotification(notification)
```

**GTNotificationDelegate Methods**
```swift
// MARK: GTNotificationDelegate Methods
    
func notificationDidDismiss(notification: GTNotification)
{
    // The notification was dismissed automatically
    NSLog("The notification was dismissed automatically")
}
    
func notificationFontForTitleLabel(notification: GTNotification) -> UIFont
{
    return UIFont(name: "AvenirNext-Medium", size: 16.0)!
}
    
func notificationFontForMessageLabel(notification: GTNotification) -> UIFont
{
    return UIFont(name: "AvenirNext-Regular", size: 13.0)!
}
```

**Screenshots**

*Dark Blur*

![Dark Blur Notification](https://raw.githubusercontent.com/MathieuWhite/GTNotification/screenshots/Screenshots/Dark%20Blur.png)

*Light Blur*

![Light Blur Notification](https://raw.githubusercontent.com/MathieuWhite/GTNotification/screenshots/Screenshots/Light%20Blur.png)

*Extra Light Blur*

![Extra Light Blur Notification](https://raw.githubusercontent.com/MathieuWhite/GTNotification/screenshots/Screenshots/Extra%20Light%20Blur.png)

*Solid Color*

![Solid Color Notification](https://raw.githubusercontent.com/MathieuWhite/GTNotification/screenshots/Screenshots/Solid%20Color.png)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
`GTNotification` adheres to [Semantic Versioning](http://semver.org/).

- `0.x` Releases - From [Release 0.1](https://github.com/MathieuWhite/GTNotification/releases/tag/0.1) To [Release 0.1](https://github.com/MathieuWhite/GTNotification/releases/tag/0.1)

---

#### [Release 0.1](https://github.com/MathieuWhite/GTNotification/releases/tag/0.1)
- GTNotification.podspec + CHANGELOG.md files added.


GTNotification/GTNotification.swift
//
//  GTNotification.swift
//  An in app notification banner for Swift.
//
//  Version 0.1
//
//  Created by Mathieu White on 2015-06-20.
//  Copyright (c) 2015 Mathieu White. All rights reserved.
//

import UIKit

@objc public protocol GTNotificationDelegate: NSObjectProtocol
{
    /**
    Tells the delegate that the notification has been dismissed
    and is no longer visible on the application's window.
    
    :param: notification the notification object that was dismissed
    */
    optional func notificationDidDismiss(notification: GTNotification)
    
    /**
    Asks the delegate for the font that should be used to display
    the title on the notification.
    
    :param: notification the object requesting a font for its title
    
    :returns: the font for the notitifcation's title
    */
    optional func notificationFontForTitleLabel(notification: GTNotification) -> UIFont
    
    /**
    Asks the delegate for the font that should be used to display
    the message on the notification.
    
    :param: notification the object requesting a font for its message
    
    :returns: the font for the notitifcation's message
    */
    optional func notificationFontForMessageLabel(notification: GTNotification) -> UIFont
}

/**
A GTNotification object specifies the properties of the
notification to display on the application's window.
*/
public class GTNotification: NSObject
{
    /// The title of the notification
    public var title: String = "Sample Notification"
    
    /// The message of the notification
    public var message: String = "This is a sample notification."
    
    /// The color of the notifiation background. If blurEnabled is true, the color will be ignored. The default color is white
    public var backgroundColor: UIColor = UIColor.whiteColor()
    
    /// The color of the text and image of the notification. The default value is black
    public var tintColor: UIColor = UIColor.blackColor()
    
    /// The image icon for the notification
    public weak var image: UIImage?
    
    /// True if the notification should blur the content it convers, false otherwise. The default value is false
    var blurEnabled: Bool = false {
        didSet {
            if (self.blurEnabled == true)
            {
                self.blurEffectStyle = UIBlurEffectStyle.Light
            }
        }
    }
    
    /// The blur effect style of the notification when blurEnabled is true. The default value is Light.
    var blurEffectStyle: UIBlurEffectStyle?
    
    /// The duration the notification should be displayed for. The default duration is 3 seconds
    var duration: NSTimeInterval = 3.0
    
    /// The position of the notification when presented on the window. The default position is Top
    var position: GTNotificationPosition = GTNotificationPosition.Top
    
    /// The animation of the notification when presented. The default animation is Fade
    public var animation: GTNotificationAnimation = GTNotificationAnimation.Fade
    
    /// The action to be performed when the notification is dismissed manually
    var action: Selector?
    
    /// The target object to which the action message is sent
    public var target: AnyObject?
    
    /// The delegate of the GTNotification
    public weak var delegate: GTNotificationDelegate?
    
    /**
    Adds a target and action for a particular event to an interal dispatch table.
    
    :param: target the target object to which the action message is sent
    :param: action a selector identifying an action message
    */
    public func addTarget(target: AnyObject, action: Selector)
    {
        self.target = target
        self.action = action
    }
}


GTNotification/GTNotificationManager.swift
//
//  GTNotificationManager.swift
//  An in app notification banner for Swift.
//
//  Version 0.1
//
//  Created by Mathieu White on 2015-06-20.
//  Copyright (c) 2015 Mathieu White. All rights reserved.
//

import UIKit

/// Identifies the position of the GTNotificationView when presented on the window
public enum GTNotificationPosition: UInt
{
    case Top
    case Bottom
}

/// Identifies the animation of the GTNotificationView when presenting
public enum GTNotificationAnimation: UInt
{
    case Fade
    case Slide
}

/**
The GTNotificationViewDelegate protocol defines an optional method to receive
user touches on objects.
*/
@objc public protocol GTNotificationViewDelegate: NSObjectProtocol
{
    /**
    Tells the delegate the notification view was tapped.
    
    :param: notificationView the GTNotificationView object being tapped
    */
    optional func notificationViewTapped(notificationView: GTNotificationView)
}

/**
A GTNotificationView object specifies a GTNotification that can be
displayed in the app on the key window.
*/
public class GTNotificationView: UIView
{
    // MARK: - Constants
    
    /// The vertical padding of the GTNotificationView
    let verticalPadding: CGFloat = 20.0
    
    /// The horizontal padding of the GTNotificationView
    let horizontalPadding: CGFloat = 20.0
    
    /// The vertical spacing between the title and message labels
    let labelSpacing: CGFloat = 16.0
    
    // MARK: - Variables
    
    /// Identifies if the GTNotificationView is visible or not
    var isVisible: Bool = false
    
    /// The title label of the GTNotificationView
    weak var titleLabel: UILabel?
    
    /// The message label of the GTNotificationView
    weak var messageLabel: UILabel?
    
    /// The image view of the GTNotificationView
    weak var imageView: UIImageView?
    
    override public var tintColor: UIColor? {
        didSet
        {
            self.titleLabel?.textColor = tintColor
            self.messageLabel?.textColor = tintColor
            self.imageView?.tintColor = tintColor
        }
    }
    
    /// The position of the GTNotificationView
    var position: GTNotificationPosition = GTNotificationPosition.Top
    
    /// The animation of the GTNotificationView
    public var animation: GTNotificationAnimation = GTNotificationAnimation.Fade
    
    /// The height of the GTNotificationView
    var notificationViewHeight: CGFloat?
    
    /// The visual blur effect for the notification view
    var blurView: UIVisualEffectView?
    
    /// The top layout constraint of the GTNotificationView
    var topConstraint: NSLayoutConstraint?
    
    /// The bottom layout constraint of the GTNotificationView
    var bottomConstraint: NSLayoutConstraint?
    
    /// The tap gesture that fires the action of the notification
    var tapGesture: UITapGestureRecognizer?
    
    /// The delegate of the GTNotificationView
    public weak var delegate: GTNotificationViewDelegate?
    
    // MARK: - Initialization
    
    required public init(coder aDecoder: NSCoder)
    {
        super.init(coder: aDecoder)
    }
    
    override init(frame: CGRect)
    {
        super.init(frame: frame)
        
        self.initNotificationView()
    }
    
    convenience init()
    {
        self.init(frame: CGRectZero)
    }
    
    func initNotificationView()
    {
        // Set the default properties of the GTNotificationView
        self.backgroundColor = UIColor.clearColor()
        self.alpha = 1.0
        self.setTranslatesAutoresizingMaskIntoConstraints(false)
        
        // Initialize the title label for the notification view
        let titleLabel: UILabel = UILabel()
        titleLabel.font = UIFont.systemFontOfSize(16.0)
        titleLabel.setTranslatesAutoresizingMaskIntoConstraints(false)
        
        // Initialize the message label for the notification view
        let messageLabel: UILabel = UILabel()
        messageLabel.font = UIFont.systemFontOfSize(13.0)
        messageLabel.numberOfLines = 0
        messageLabel.lineBreakMode = NSLineBreakMode.ByWordWrapping
        messageLabel.setTranslatesAutoresizingMaskIntoConstraints(false)
        
        // Initialize the image view for the notification view
        let imageView: UIImageView = UIImageView()
        imageView.setTranslatesAutoresizingMaskIntoConstraints(false)
        
        // Initialize the tap gesture
        let tapGesture: UITapGestureRecognizer = UITapGestureRecognizer(target: self, action: Selector("notificationTapped"))
        self.addGestureRecognizer(tapGesture)
        
        // Add the labels to the notification view
        self.addSubview(titleLabel)
        self.addSubview(messageLabel)
        self.addSubview(imageView)
        
        // Set each label to their variable
        self.titleLabel = titleLabel
        self.messageLabel = messageLabel
        self.imageView = imageView
        self.tapGesture = tapGesture
    }
    
    // MARK: - Auto Layout
    
    /**
    This method will setup the constraints for the notification view.
    */
    func setupConstraints()
    {
        if (self.blurView != nil)
        {
            // Layout the blur view vertically
            self.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("V:|[_blur]|",
                options: NSLayoutFormatOptions(0),
                metrics: nil,
                views: ["_blur" : self.blurView!]))
            
            // Layout the blur view horizontally
            self.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("H:|[_blur]|",
                options: NSLayoutFormatOptions(0),
                metrics: nil,
                views: ["_blur" : self.blurView!]))
        }
        
        if (self.imageView?.image != nil)
        {
            // Image Center Y
            self.addConstraint(NSLayoutConstraint(item: self.imageView!,
                attribute: NSLayoutAttribute.CenterY,
                relatedBy: NSLayoutRelation.Equal,
                toItem: self,
                attribute: NSLayoutAttribute.CenterY,
                multiplier: 1.0,
                constant: 0.0))
            
            // Image Left
            self.addConstraint(NSLayoutConstraint(item: self.imageView!,
                attribute: NSLayoutAttribute.Left,
                relatedBy: NSLayoutRelation.Equal,
                toItem: self,
                attribute: NSLayoutAttribute.Left,
                multiplier: 1.0,
                constant: self.horizontalPadding))
            
            // Image Width
            self.addConstraint(NSLayoutConstraint(item: self.imageView!,
                attribute: NSLayoutAttribute.Width,
                relatedBy: NSLayoutRelation.Equal,
                toItem: nil,
                attribute: NSLayoutAttribute.NotAnAttribute,
                multiplier: 1.0,
                constant: 36.0))
            
            // Image Height
            self.addConstraint(NSLayoutConstraint(item: self.imageView!,
                attribute: NSLayoutAttribute.Height,
                relatedBy: NSLayoutRelation.Equal,
                toItem: nil,
                attribute: NSLayoutAttribute.NotAnAttribute,
                multiplier: 1.0,
                constant: 36.0))
        }
        
        // Title Label Top
        self.addConstraint(NSLayoutConstraint(item: self.titleLabel!,
            attribute: NSLayoutAttribute.Top,
            relatedBy: NSLayoutRelation.Equal,
            toItem: self,
            attribute: NSLayoutAttribute.Top,
            multiplier: 1.0,
            constant: self.verticalPadding))
        
        // Title Label Left
        self.addConstraint(NSLayoutConstraint(item: self.titleLabel!,
            attribute: NSLayoutAttribute.Left,
            relatedBy: NSLayoutRelation.Equal,
            toItem: (self.imageView?.image == nil) ? self : self.imageView!,
            attribute: (self.imageView?.image == nil) ? NSLayoutAttribute.Left : NSLayoutAttribute.Right,
            multiplier: 1.0,
            constant: self.horizontalPadding))
        
        // Title Label Right
        self.addConstraint(NSLayoutConstraint(item: self.titleLabel!,
            attribute: NSLayoutAttribute.Right,
            relatedBy: NSLayoutRelation.Equal,
            toItem: self,
            attribute: NSLayoutAttribute.Right,
            multiplier: 1.0,
            constant: -self.horizontalPadding))
        
        // Message Label Top
        self.addConstraint(NSLayoutConstraint(item: self.messageLabel!,
            attribute: NSLayoutAttribute.Top,
            relatedBy: NSLayoutRelation.Equal,
            toItem: self.titleLabel!,
            attribute: NSLayoutAttribute.Baseline,
            multiplier: 1.0,
            constant: 0.0))
        
        // Message Label Left
        self.addConstraint(NSLayoutConstraint(item: self.messageLabel!,
            attribute: NSLayoutAttribute.Left,
            relatedBy: NSLayoutRelation.Equal,
            toItem: self.titleLabel!,
            attribute: NSLayoutAttribute.Left,
            multiplier: 1.0,
            constant: 0.0))
        
        // Message Label Right
        self.addConstraint(NSLayoutConstraint(item: self.messageLabel!,
            attribute: NSLayoutAttribute.Right,
            relatedBy: NSLayoutRelation.Equal,
            toItem: self.titleLabel!,
            attribute: NSLayoutAttribute.Right,
            multiplier: 1.0,
            constant: 0.0))
        
        // Message Label Bottom
        self.addConstraint(NSLayoutConstraint(item: self.messageLabel!,
            attribute: NSLayoutAttribute.Bottom,
            relatedBy: NSLayoutRelation.Equal,
            toItem: self,
            attribute: NSLayoutAttribute.Bottom,
            multiplier: 1.0,
            constant: -self.verticalPadding))
    }
    
    /**
    This method will layout the notification view and add it to the key window of the application.
    */
    func layoutNotificationViewInWindow()
    {
        // Get a reference of the application window
        var window: UIWindow? = UIApplication.sharedApplication().keyWindow
        
        // The application has a key window
        if let window = window
        {
            // Add the notification view to the window
            window.addSubview(self)
            
            // Calculate the height of the notification view
            self.notificationViewHeight = self.heightForNoticationView()
            
            // Views Dictionary
            var viewsDict: [NSObject : AnyObject] = [NSObject : AnyObject]()
            viewsDict["_view"] = self

            // Metrics Dictionary
            var metricsDict: [NSObject : AnyObject] = [NSObject : AnyObject]()
            metricsDict["_h"] = self.notificationViewHeight
            
            // Notification View Width
            let notificationHorizontalConstraints: [AnyObject] = NSLayoutConstraint.constraintsWithVisualFormat("H:|[_view]|",
                options: NSLayoutFormatOptions(0),
                metrics: metricsDict,
                views: viewsDict)
            
            // Notification View Height
            let notificationVerticalConstraints: [AnyObject] = NSLayoutConstraint.constraintsWithVisualFormat("V:[_view(_h)]",
                options: NSLayoutFormatOptions(0),
                metrics: metricsDict,
                views: viewsDict)
            
            // Notification View Top
            if (self.position == GTNotificationPosition.Top)
            {
                var topConstant: CGFloat = 0.0

                if (self.animation == GTNotificationAnimation.Slide)
                {
                    topConstant = -self.notificationViewHeight!
                }
                
                let topConstraint: NSLayoutConstraint = NSLayoutConstraint(item: self,
                    attribute: NSLayoutAttribute.Top,
                    relatedBy: NSLayoutRelation.Equal,
                    toItem: window,
                    attribute: NSLayoutAttribute.Top,
                    multiplier: 1.0,
                    constant: topConstant)
                
                window.addConstraint(topConstraint)
                self.topConstraint = topConstraint
            }
            
            // Notification View Bottom
            if (self.position == GTNotificationPosition.Bottom)
            {
                var bottomConstant: CGFloat = 0.0
                
                if (self.animation == GTNotificationAnimation.Slide)
                {
                    bottomConstant = self.notificationViewHeight!
                }
                
                let bottomConstraint: NSLayoutConstraint = NSLayoutConstraint(item: self,
                    attribute: NSLayoutAttribute.Bottom,
                    relatedBy: NSLayoutRelation.Equal,
                    toItem: window,
                    attribute: NSLayoutAttribute.Bottom,
                    multiplier: 1.0,
                    constant: bottomConstant)
                
                window.addConstraint(bottomConstraint)
                self.bottomConstraint = bottomConstraint
            }
            
            // Add the constraints to the window
            window.addConstraints(notificationVerticalConstraints)
            window.addConstraints(notificationHorizontalConstraints)
            
            // Layout the notification view
            self.layoutIfNeeded()
            self.titleLabel?.layoutIfNeeded()
            self.messageLabel?.layoutIfNeeded()
        }
        else
        {
            // No key window found
            assertionFailure("Warning: make sure to call makeKeyAndVisible on the application's window.")
        }
    }
    
    // MARK: - Instance Methods
    
    /**
    This method calculates and returns the height for the notification view.
    
    :returns: the height of the notification view
    */
    func heightForNoticationView() -> CGFloat
    {
        // Determine the maximum with of our labels
        let maximumLabelWidth: CGFloat = CGRectGetWidth(UIScreen.mainScreen().bounds) - (self.horizontalPadding * 2.0)
        
        // Initialize our maximum label size
        let maximumLabelSize: CGSize = CGSizeMake(maximumLabelWidth, CGFloat.max)
        
        // Get the height of the title label
        let titleLabelHeight: CGFloat = (self.titleLabel!.text! as NSString).boundingRectWithSize(maximumLabelSize,
            options: NSStringDrawingOptions.UsesLineFragmentOrigin,
            attributes: [NSFontAttributeName : self.titleLabel!.font],
            context: nil).height
        
        // Get the height of the message label
        let messageLabelHeight: CGFloat = (self.messageLabel!.text! as NSString).boundingRectWithSize(maximumLabelSize,
            options: NSStringDrawingOptions.UsesLineFragmentOrigin,
            attributes: [NSFontAttributeName : self.messageLabel!.font],
            context: nil).height
        
        // Return the total height, (vertical top and bottom padding + label heights + label spacing)
        return self.verticalPadding * 2.0 + titleLabelHeight + messageLabelHeight + self.labelSpacing
    }
    
    /**
    This method applies the notification view's visual attributes.
    
    :param: notification the notification that will be displayed
    */
    func prepareForNotification(notification: GTNotification)
    {
        self.tintColor = notification.tintColor
        self.titleLabel?.text = notification.title
        self.messageLabel?.text = notification.message
        self.imageView?.image = notification.image
        self.position = notification.position
        self.animation = notification.animation
        
        // Get the font for the title label
        if (notification.delegate?.respondsToSelector(Selector("notificationFontForTitleLabel:")) == true)
        {
            self.titleLabel?.font = notification.delegate!.notificationFontForTitleLabel!(notification)
        }
        
        // Get the font for the message label
        if (notification.delegate?.respondsToSelector(Selector("notificationFontForMessageLabel:")) == true)
        {
            self.messageLabel?.font = notification.delegate!.notificationFontForMessageLabel!(notification)
        }
        
        if (notification.blurEnabled == true)
        {
            self.backgroundColor = UIColor.clearColor()
            
            // Add the blur effect to the notification view
            let blurEffect: UIBlurEffect = UIBlurEffect(style: notification.blurEffectStyle!)
            let blurView: UIVisualEffectView = UIVisualEffectView(effect: blurEffect)
            blurView.setTranslatesAutoresizingMaskIntoConstraints(false)
            self.insertSubview(blurView, atIndex: 0)
            self.blurView = blurView
        }
        else
        {
            self.backgroundColor = notification.backgroundColor
        }
        
        // Set the initial alpha to 0.0 for the fade animation
        if (self.animation == GTNotificationAnimation.Fade)
        {
            self.alpha = 0.0
        }
        else
        {
            self.alpha = 1.0
        }
        
        // Layout the view's subviews
        self.setupConstraints()
        
        // Layout the view in the window
        self.layoutNotificationViewInWindow()
    }
    
    /**
    This method animates the notification view on the application's window
    
    :param: show         true if the notification view will dismiss, false otherwise
    :param: completion   the completion closure to execute after the animation
    */
    func animateNotification(willShow show: Bool, completion: (finished: Bool) -> Void)
    {
        // Set the animation view's visibility
        self.isVisible = show
        
        // The notification view should animate with a fade
        if (self.animation == GTNotificationAnimation.Fade)
        {
            UIView.animateWithDuration(0.4,
                animations: {
                    self.alpha = (show == true) ? 1.0 : 0.0
                },
                completion: { (finished) -> Void in
                    completion(finished: finished)
            })
        }
        
        // The notification view should animate with a slide
        if (self.animation == GTNotificationAnimation.Slide)
        {
            if (self.position == GTNotificationPosition.Top)
            {
                self.topConstraint?.constant = (show == true) ? 0.0 : -self.notificationViewHeight!
            }
            else
            {
                self.bottomConstraint?.constant = (show == true) ? 0.0 : self.notificationViewHeight!
            }
            
            UIView.animateWithDuration(0.4,
                animations: {
                    self.layoutIfNeeded()
                },
                completion: {(finished) -> Void in
                    completion(finished: finished)
            })
        }
    }
    
    // MARK: - Gesture Recognizer Methods
    
    /**
    This method fires when the notification view is tapped. 
    
    NOTE: Tapping the notification view will dismiss it immediately.
    */
    @objc private func notificationTapped()
    {
        // Notify the delegate that the notification view was tapped
        if (self.delegate?.respondsToSelector(Selector("notificationViewTapped:")) == true)
        {
            self.delegate!.notificationViewTapped!(self)
        }
    }
}

public class GTNotificationManager: NSObject, GTNotificationViewDelegate
{
    // MARK: - Variables
    
    /// The singleton instance of the GTNotificationManager
    public static var sharedInstance: GTNotificationManager = GTNotificationManager()
    
    /// The private array of notifications queued for display
    private var mutableNotifications: [GTNotification] = []
    
    /// The current notification view being displayed
    private var currentNotificationView: GTNotificationView?
    
    /// The timer that fires the scheduled dismissal of the notification
    private var dismissalTimer: NSTimer?
    
    // MARK: - Read Only
    
    /// The array of notifications queued for display
    internal var notifications: [GTNotification] {
        get {
            let immutableNotifications = self.mutableNotifications
            return immutableNotifications
        }
    }
    
    // MARK: - Initializers
    
    override init()
    {
        super.init()
    }
    
    // MARK: - Instance Methods
    
    /**
    This method shows the notification on the application's window.
    
    :param: notification the notification to display
    */
    public func showNotification(notification: GTNotification)
    {
        // Only show one notification at a time
        if (self.currentNotificationView == nil)
        {
            // Queue the notification for display
            self.mutableNotifications.append(notification)
            
            // Initialize a GTNotificationView
            let notificationView: GTNotificationView = GTNotificationView()
            notificationView.delegate = self
            
            // Prepare the view for the notification to display
            notificationView.prepareForNotification(notification)
            
            // Animate the notification
            notificationView.animateNotification(willShow: true, completion: {(finished: Bool) -> Void in
                
                // Schedule the notification view's dismissal
                self.dismissalTimer = NSTimer.scheduledTimerWithTimeInterval(notification.duration,
                    target: self,
                    selector: Selector("dismissCurrentNotification:"),
                    userInfo: notification,
                    repeats: false)
                
            })
            
            // Set the current notification
            self.currentNotificationView = notificationView
        }
    }
    
    /**
    This method dismissed the current notification on the application's window.
    */
    @objc private func dismissCurrentNotification(timer: NSTimer?)
    {
        var notification: GTNotification? = timer?.userInfo as? GTNotification
        
        self.dismissalTimer?.invalidate()
        self.dismissalTimer = nil
        
        if let notificationView = self.currentNotificationView
        {
            // Animate the notification
            notificationView.animateNotification(willShow: false, completion: {(finished: Bool) -> Void in
                
                // Notify the delegate of the notification's dismissal
                if let notification = notification
                {                    
                    if (notification.delegate?.respondsToSelector(Selector("notificationDidDismiss:")) == true)
                    {
                        notification.delegate!.notificationDidDismiss!(notification)
                    }
                }
                
                notificationView.removeFromSuperview()
                self.currentNotificationView = nil
                
            })
        }
    }
    
    // MARK: - GTNotificationViewDelegate Methods
    
    @objc public func notificationViewTapped(notificationView: GTNotificationView)
    {
        let notification = self.mutableNotifications.removeAtIndex(0)
        
        self.dismissCurrentNotification(nil)
        
        if
            let target: AnyObject = notification.target,
            let action = notification.action
        {
            if (target.respondsToSelector(action) == true)
            {
                dispatch_after(DISPATCH_TIME_NOW, dispatch_get_main_queue(), {
                    NSThread.detachNewThreadSelector(action, toTarget: target, withObject: nil)
                })
            }
        }
    }
}


SampleNotifcations/SampleNotifcationsTests/SampleNotifcationsTests.swift
//
//  SampleNotifcationsTests.swift
//  SampleNotifcationsTests
//
//  Created by Mathieu White on 2015-06-21.
//  Copyright (c) 2015 Mathieu White. All rights reserved.
//

import UIKit
import XCTest

class SampleNotifcationsTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        XCTAssert(true, "Pass")
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measureBlock() {
            // Put the code you want to measure the time of here.
        }
    }
    
}


SampleNotifcations/SampleNotifcations/ViewController.swift
//
//  ViewController.swift
//  SampleNotifcations
//
//  Created by Mathieu White on 2015-06-21.
//  Copyright (c) 2015 Mathieu White. All rights reserved.
//

import UIKit

class ViewController: UIViewController, GTNotificationDelegate
{
    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        self.view.backgroundColor = UIColor.whiteColor()
        
        // Notification Button
        let button: UIButton = UIButton.buttonWithType(UIButtonType.System) as! UIButton
        button.setTitle("Show Notification", forState: UIControlState.Normal)
        button.setTranslatesAutoresizingMaskIntoConstraints(false)
        button.addTarget(self, action: Selector("notificationButtonPressed"), forControlEvents: UIControlEvents.TouchUpInside)
        
        self.view.addSubview(button)
        
        // Center the button vertically
        self.view.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("H:[_super]-(<=1)-[_btn]",
            options: NSLayoutFormatOptions.AlignAllCenterY,
            metrics: nil,
            views: ["_super" : self.view, "_btn" : button]))
        
        // Center the button horizontally
        self.view.addConstraints(NSLayoutConstraint.constraintsWithVisualFormat("V:[_super]-(<=1)-[_btn]",
            options: NSLayoutFormatOptions.AlignAllCenterX,
            metrics: nil,
            views: ["_super" : self.view, "_btn" : button]))
    }

    override func didReceiveMemoryWarning()
    {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


    func notificationButtonPressed()
    {
        // Initialize a notification
        let notification: GTNotification = GTNotification()
        notification.image = UIImage(named: "notification")?.imageWithRenderingMode(UIImageRenderingMode.AlwaysTemplate)
        notification.blurEnabled = true
        notification.blurEffectStyle = UIBlurEffectStyle.Dark
        notification.tintColor = UIColor.whiteColor()
        notification.animation = GTNotificationAnimation.Slide
        notification.delegate = self
        
        // Add a selector to perform on tap
        notification.addTarget(self, action: Selector("dismissNotification"))
        
        // Show the notification
        GTNotificationManager.sharedInstance.showNotification(notification)
    }
    
    func dismissNotification()
    {
        NSLog("The notification was dismissed manually")
        
        let alertController: UIAlertController = UIAlertController(title: "",
            message: "Thank you for checking out GTNotification.",
            preferredStyle: UIAlertControllerStyle.Alert)
        
        let dismissAction: UIAlertAction = UIAlertAction(title: "Dismiss",
            style: UIAlertActionStyle.Default) { (alert) -> Void in
                alertController.dismissViewControllerAnimated(true, completion: nil)
        }
        
        alertController.addAction(dismissAction)
        
        self.presentViewController(alertController, animated: true, completion: nil)
    }
    
    // MARK: GTNotificationDelegate Methods
    
    func notificationDidDismiss(notification: GTNotification)
    {
        // The notification was dismissed automatically
        NSLog("The notification was dismissed automatically")
    }
    
    func notificationFontForTitleLabel(notification: GTNotification) -> UIFont
    {
        return UIFont(name: "AvenirNext-Medium", size: 16.0)!
    }
    
    func notificationFontForMessageLabel(notification: GTNotification) -> UIFont
    {
        return UIFont(name: "AvenirNext-Regular", size: 13.0)!
    }
    
}
